sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/options/PipelineOptionsReflector.java

```diff
@@ -0,0 +1,179 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.dataflow.sdk.options;
+
+import com.google.cloud.dataflow.sdk.util.common.ReflectHelpers;
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+
+import java.beans.Introspector;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Utilities to reflect over {@link PipelineOptions}.
+ */
+class PipelineOptionsReflector {
+  private PipelineOptionsReflector() {}
+
+  /**
+   * Retrieve the full set of pipeline option properties visible within the type hierarchy closure
+   * for the set of input interfaces. An option is "visible" if:
+   *
+   * <ul>
+   *   <li>The option is defined within the interface hierarchy closure of the input
+   *   {@link PipelineOptions}.</li>
+   *   <li>The defining interface is not marked {@link Hidden}.</li>
+   * </ul>
+   */
+  static Set<Property> collectVisibleProperties(
+      Class<? extends PipelineOptions> optionsInterface) {
+    Iterable<Method> methods = ReflectHelpers.getClosureOfMethodsOnInterface(optionsInterface);
+    Multimap<String, Method> propsToGetters = getPropertyNamesToGetters(methods);
+
+    ImmutableSet.Builder<Property> setBuilder = ImmutableSet.builder();
+    for (Map.Entry<String, Method> propAndGetter : propsToGetters.entries()) {
+      String prop = propAndGetter.getKey();
+      Method getter = propAndGetter.getValue();
+      Class<?> declaringClass = getter.getDeclaringClass();
+
+      if (!PipelineOptions.class.isAssignableFrom(declaringClass)) {
+        continue;
+      }
+
+      if (declaringClass.isAnnotationPresent(Hidden.class)) {
+        continue;
+      }
+
+      setBuilder.add(Property.of((Class<? extends PipelineOptions>) declaringClass, prop, getter));
+    }
+
+    return setBuilder.build();
+  }
+
+  /**
+   * Retrieve the full set of pipeline option properties visible within the type hierarchy of the
+   * input interfaces. An option is "visible" if:
+   *
+   * <ul>
+   *   <li>The option is defined within the interface hierarchy closure of the input
+   *   {@link PipelineOptions}.</li>
+   *   <li>The defining interface is not marked {@link Hidden}.</li>
+   * </ul>
+   */
+  static Set<Property> collectVisibleProperties(
+      Collection<Class<? extends PipelineOptions>> optionsInterfaces) {
+    ImmutableSet.Builder<Property> setBulder = ImmutableSet.builder();
+    for (Class<? extends PipelineOptions> optionsInterface : optionsInterfaces) {
+      setBulder.addAll(collectVisibleProperties(optionsInterface));
+    }
+
+    return setBulder.build();
+  }
+
+  /**
+   * Extra properties and their respective getter methods from a series of {@link Method methods}.
```

Comment:
Make it clearer that its a map from property name to getter methods

Might want to say why there can be multiple.